Remove `unwrap` from transforms and add `ungroup` to more blocks

ntsekouras opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Part of: https://github.com/WordPress/gutenberg/issues/45841

Regarding the above issue, this PR will not add any shortcuts to reduce its scope. Additionally this PR could (and maybe will) be split in two separate ones. This is because it does two things:
1. Remove the `unwrap` from the block switcher for `Group, Columns, Quote` blocks and essentially partially or fully reverting [1](https://github.com/WordPress/gutenberg/pull/45666/), [2](https://github.com/WordPress/gutenberg/pull/42685), [3](https://github.com/WordPress/gutenberg/pull/39718). 
2. Adds a new API for blocks to be able to `ungroup`


## How and Why?
This PR suggests a new `transforms` API for the `ungroup` functionality in block settings menu controls(ellipsis menu in block toolbar). If a block, besides the set grouping block(see [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/blocks/README.md#setgroupingblockname)), want to render the `Ungroup` button, should add an `ungroup` function in their `transforms` of the block type.

This new function accepts the block's attributes and innerBlocks(although we might consider passing the whole block) and is responsible for returning the new blocks to replace the selected block. This is because blocks can have a different unwrapping handling - ex. Columns need to also remove the `core/column` blocks etc..

**Why in `transforms` API?**
No strong opinions here, but I felt it's less impactful than adding a new `Block` API and it can make sense semantically. We could do it there too though.

**Why we even need a new API?**
My first try was just adding the `Ungroup` button inside the blocks' edit function with `BlockSettingsMenuControls`. That would be the best option, if we could handle the wanted position of the menu item. Right now the `fills` are rendered above the `Group` menu item and creates a disconnect with the rest of the menu items. 

## Testing Instructions
1. Insert `Group, Quote, Columns` blocks
2. Observe the `unwrap` option in block switcher is gone
3. Observe that the `Ungroup` menu item in the ellipsis menu in block toolbar works as the `unwrap` button was working. I'm saying this because there is already a flow there in the logic about when to show this button. Checking innerBlocks existence is not enough because for example in an seemingly empty Columns block we have empty `core/column` blocks.
4. Observe that transform options in block switcher are not affected besides the removal of `unwrap` option.

### Tasks
- [x] Agreement on API/approach
- [x] Add docs
- [x] Add/update tests
- [x] Check with mobile as I saw some tests with `unwrap`

ntsekouras committed:
Remove `unwrap` from transforms and add `ungroup` to more blocks

jasmussen commented:
Thanks for this one! 

<img width="669" alt="group with ellipsis menu open, and focus on the ungroup option" src="https://user-images.githubusercontent.com/1204802/236783388-ddfb51f2-31b7-48b1-85b9-0f2ecd9c5792.png">

Designwise I would suggest this one is good to go, for a few reasons:

* The transform interface should be a focused overview of other blocks and patterns you can transform to, and ideally not a place for options like "Unwrap".
* Group already exists in the ellipsis menu, as does "Ungroup".
* "Unwrap" is not an easy to understand term, and the icon that went with it makes it look like you transform into a group. "Ungroup" is a simpler term, even when it technically unwraps a _columns_ block, which you can still think of as just a group of two columns.

We should follow up with ⌘G and ⌘⇧G for grouping and ungrouping later on.

gziolo commented:
This is an interesting idea to extend `transforms` with a new type `ungroup`. I'm resharing the same idea in https://github.com/WordPress/gutenberg/pull/49464#issuecomment-1538054025 in the context of exposing a new way to modify the block when copying it.

richtabor reviewed (approved):
Works as expected, nice.

richtabor commented:
> The transform interface should be a focused overview of other blocks and patterns you can transform to, and ideally not a place for options like "Unwrap".

I agree. Though it begs to question if transforming to "Group" or "Columns" should be a transform. They're not _really_ transforming. And we already have "Group" as an action within the Block Settings Menu (same as Ungroup).

ntsekouras committed:
add docs

jasmussen commented:
> I agree. Though it begs to question if transforming to "Group" or "Columns" should be a transform. They're not really transforming. And we already have "Group" as an action within the Block Settings Menu (same as Ungroup).

Fuzzy separate conversation, yes. I think it can be valid as a transform: _transforming_ a columns block should essentially change the container into a _group_ instead. Whereas if you choose the "Group" item from the ellipsis menu, it should _wrap_ the columns block in a group. That's my instinct: transforms should always transform.

Multi-selection falls under that umbrella, select 3 paragraphs and transform to group or columns, and that should work as it does today. 

The main thing I personally find confusing, is when the option listed as a transform actually _wraps_ in a new container instead of actually transforming.

ntsekouras committed:
update e2e tests

ntsekouras committed:
add e2e test

mcsf commented on docs/reference-guides/block-api/block-transforms.md:
```diff
@@ -270,7 +270,7 @@ A transformation of type `shortcode` is an object that takes the following param
 -   **type** _(string)_: the value `shortcode`.
 -   **tag** _(string|array)_: the shortcode tag or list of shortcode aliases this transform can work with.
--   **transform** _(function, optional): a callback that receives the shortcode attributes as the first argument and the [WPShortcodeMatch](/packages/shortcode/README.md#next) as the second. It should return a block object or an array of block objects. When this parameter is defined, it will take precedence over the `attributes` parameter.
+-   **transform** \_(function, optional): a callback that receives the shortcode attributes as the first argument and the [WPShortcodeMatch](/packages/shortcode/README.md#next) as the second. It should return a block object or an array of block objects. When this parameter is defined, it will take precedence over the `attributes` parameter.
```
```suggestion
-   **transform** _(function, optional)_: a callback that receives the shortcode attributes as the first argument and the [WPShortcodeMatch](/packages/shortcode/README.md#next) as the second. It should return a block object or an array of block objects. When this parameter is defined, it will take precedence over the `attributes` parameter.
```

mcsf commented on docs/reference-guides/block-api/block-transforms.md:
```diff
@@ -336,3 +336,25 @@ transforms: {
+
+## `ungroup` blocks
+
+Via the optional `transforms` key of the block configuration, blocks can use the `ungroup` subkey to define which inner blocks of a grouping block can be extracted, replacing the block.
```
It's not just/necessarily inner blocks: attributes may be mined as well, as we see with the Quote block's `citation` attribute.

ntsekouras replied:
I updated a bit the wording here: https://github.com/WordPress/gutenberg/pull/50385/commits/2edc576a6c574449ed797edd76f6d8f7fb3bb832.

mcsf commented on docs/reference-guides/block-api/block-transforms.md:
```diff
@@ -336,3 +336,25 @@ transforms: {
+
+Via the optional `transforms` key of the block configuration, blocks can use the `ungroup` subkey to define which inner blocks of a grouping block can be extracted, replacing the block.
+
+If a block has an `ungroup` transform, it is considered ungroupable, without the requirement of being the default grouping block. The UI used to ungroup a block with this API, is the same with the one used for the default grouping block. In order for the ugroup UI to render, we must have a single grouping block selected, which also contains some inner blocks.
```
```suggestion
If a block has an `ungroup` transform, it is considered ungroupable, without the requirement of being the default grouping block. The UI used to ungroup a block with this API is the same as the one used for the default grouping block. In order for the Ungroup button to be displayed, we must have a single grouping block selected, which also contains some inner blocks.
```

fluiddot commented:
Hey @ntsekouras 👋, as we talked internally regarding the mobile native version, it would be great to postpone merging this PR until we implement the mobile native version of [`ConvertToGroupButton` component](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/convert-to-group-buttons/index.native.js). This is needed to provide the Group/Ungroup buttons in the app, otherwise, these options won't be available in the app.

I think the implementation shouldn't be complex if we can use the `useConvertToGroupButtonProps` hook. Although, the way [the block menu actions](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-mobile-toolbar/block-actions-menu.native.js) are rendered in the app differs from the web. This part is what might take most of the effort.

I'm afraid I don't bandwidth this week to focus on this, but hopefully, I can spare time next week and unblock the mobile part, as well as address the failure in unit tests. Thanks 🙇  !

mcsf committed:
Refactor useConvertToGroupButtonProps for readability

mcsf commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -2118,22 +2116,10 @@ export const getBlockTransformItems = createSelector(
 		);
 
-		// Consider unwraping the highest priority.
-		itemsByName[ '*' ] = {
```
Good to see these quirks gone :)

mcsf commented on packages/block-editor/src/components/convert-to-group-buttons/use-convert-to-group-button-props.js:
```diff
@@ -31,68 +31,62 @@ import { store as blockEditorStore } from '../../store';
-		blocksSelection,
-		groupingBlockName,
-	} = useSelect(
+	return useSelect(
```
In 9a2fb4a7ec607599b633424ee95ac3e7fe32980f I took the liberty of refactoring this hook, which was IMO adding unnecessary noise and was repacking objects for no reason:

```js
const { a, b, c } = useSelect( ( select ) => {
  const _a = ...
  const _b = ...
  const _c = ...
  return { a: _a, b: _b, c: _c }
} )
return { a, b, c }
```

becomes

```js
return useSelect( ( select ) => {
  const a = ...
  const b = ...
  const c = ...
  return { a, b, c }
} )
```

ntsekouras replied:
I didn't notice that. That's better, thank you!

mcsf reviewed (commented):
This looks great, I have no objections!

ntsekouras committed:
Apply suggestions from code review

Co-authored-by: Miguel Fonseca <miguelcsf@gmail.com>

ntsekouras committed:
update docs

gziolo commented on docs/reference-guides/block-api/block-transforms.md:
```diff
@@ -341,7 +341,7 @@ transforms: {
 Via the optional `transforms` key of the block configuration, blocks can use the `ungroup` subkey to define which inner blocks of a grouping block can be extracted, replacing the block.
 
-If a block has an `ungroup` transform, it is considered ungroupable, without the requirement of being the default grouping block. The UI used to ungroup a block with this API, is the same with the one used for the default grouping block. In order for the ugroup UI to render, we must have a single grouping block selected, which also contains some inner blocks.
+If a block has an `ungroup` transform, it is considered ungroupable, without the requirement of being the default grouping block. The UI used to ungroup a block with this API is the same as the one used for the default grouping block. In order for the Ungroup button to be displayed, we must have a single grouping block selected, which also contains some inner blocks.
```
> it is considered ungroupable

I think this phrase might be confusing for folks because "ungroupable" means "Incapable of being grouped together".

mcsf replied:
Agreed. Depending on the use in a sentence, some ideas: "able to be ungrouped", "ungroup-ready", "eligible for ungrouping"? 😅

ntsekouras replied:
You're right. I went for `eligible for ungrouping`.

ntsekouras committed:
reword ungrouping in docs

gziolo reviewed (commented):
This PR is shaping up nicely. Great work @ntsekouras 👍🏻

kevin940726 commented on test/e2e/specs/editor/blocks/columns.spec.js:
```diff
@@ -82,4 +82,54 @@ test.describe( 'Columns', () => {
+<!-- wp:paragraph -->
+<p>2</p>
+<!-- /wp:paragraph -->`
+		);
```
Prefer `getBlocks` as what we discussed in the other PR :)

ntsekouras committed:
update e2e test

fluiddot commented:
> Hey @ntsekouras 👋, as we talked internally regarding the mobile native version, it would be great to postpone merging this PR until we implement the mobile native version of [`ConvertToGroupButton` component](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/convert-to-group-buttons/index.native.js). This is needed to provide the Group/Ungroup buttons in the app, otherwise, these options won't be available in the app.

In https://github.com/WordPress/gutenberg/pull/50693, we'll incorporate into this PR the needed changes for the mobile native version (including unit test fixes).

fluiddot committed:
[RNMobile] Add Group and Ungroup block actions (#50693)

* Add `useConvertToGroupButtons` hook

Most of the logic of this hook has been extracted from `ConvertToGroupButton` component. The main difference is that we return the configuration for the block actions instead of a component.

* Add Group and Ungroup options to block actions menu

* Remove `canUnwrap` option from `getBlockTransformOptions` test helper

* Update tests for Group, Quote and Columns blocks

fluiddot commented:
> > Hey @ntsekouras 👋, as we talked internally regarding the mobile native version, it would be great to postpone merging this PR until we implement the mobile native version of [`ConvertToGroupButton` component](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/convert-to-group-buttons/index.native.js). This is needed to provide the Group/Ungroup buttons in the app, otherwise, these options won't be available in the app.
> 
> In #50693, we'll incorporate into this PR the needed changes for the mobile native version (including unit test fixes).

The mobile native changes have been incorporating into this PR via https://github.com/WordPress/gutenberg/pull/50385/commits/5c6d77b127ec2d41d29f48386e5cee129f8d93db.
